Update docs on building and hosting Docker images (#640)
broadinstitute/gatk-sv: website/docs/advanced/docker/manual.md

---
title: Manual Deployment
description: Build and Publish Images
sidebar_position: 3
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> If you contribute to the GATK-SV codebase, we recommend you build Docker images 
> to ensure they build successfully and function as intended. The process involves two steps:

Reviewer: ```suggestion
If you contribute to the GATK-SV codebase, we recommend you ensure that affected Docker images build successfully and function as intended. The process involves two steps:
```

> - **Build**: Create Docker images from Dockerfiles.
> 
> - **Publish**: Upload the built Docker images to container registries 
> (e.g., Google or Azure container registries, GCR and ACR, respectively) 
> to make them available for use in Terra or Cromwell.
> _You may skip this step unless you would like to host the images you built on your own container registry._

Reviewer: ```suggestion
1. **Build**: Create Docker images from Dockerfiles.

2. **Publish**: Upload the built Docker images to container registries 
(e.g., Google or Azure container registries, GCR and ACR, respectively) 
to make them available for use in Terra or Cromwell.
_You may skip this step unless you would like to host the images you built on your own container registry._
```

To streamline the process, we have developed a 
[script](https://github.com/broadinstitute/gatk-sv/blob/main/scripts/docker/build_docker.py)
that automates both the build and publish steps. 
This section provides guidelines on setting up the environment and running the 
script with a minimal example.

> :::danger Linux Machine Required
> Only Linux machines (dedicated or virtual) are supported for building GATK-SV Docker images. 
> Images created on non-Linux machines (e.g., Apple M1) may not function as intended, 
> even if the build process runs successfully. 
> :::

Reviewer: ```suggestion
:::danger x86 Linux Machine Required
Only Linux machines (dedicated or virtual) are supported for building GATK-SV Docker images. 
In addition, images created on non-Intel processor architectures (e.g., Apple M1) may not function as intended, 
even if the build process runs successfully. 
:::
```

on a cloud service provider.
You may [skip to the next section](#checkout) if you are using a dedicated Linux machine 
(e.g., a laptop running Ubuntu).

#### 1. Set environment variables

<Tabs
 groupId="cloud"
 defaultValue="gcp"
 values={[
  { label: 'GCP', value: 'gcp', }
 ]
}>
 <TabItem value="gcp">
    ```bash
    export PROJECT_ID="<GOOGLE PROJECT ID>"
    export ZONE_ID="<ZONE ID>"
    
    # Make sure no machine with the following name exist,
    # and you follow VM naming conventions, e.g., all lower-case characters.
    export INSTANCE_NAMES="<VM NAME>"
    ```

 </TabItem>
</Tabs>

#### 2. Create an Ubuntu VM 
You may [skip to the next step](#connect-to-vm) if you have already created a VM.

<Tabs
 groupId="cloud"
 defaultValue="gcp"
 values={[
  { label: 'GCP', value: 'gcp', }
 ]
}>
 <TabItem value="gcp">
    ```bash
    gcloud compute instances create $INSTANCE_NAMES \
      --project=$PROJECT_ID \
      --zone=$ZONE_ID \
      --machine-type=e2-standard-2 \
      --create-disk=auto-delete=yes,boot=yes,device-name=$INSTANCE_NAMES,image=projects/ubuntu-os-cloud/global/images/ubuntu-2310-mantic-amd64-v20240213,mode=rw,size=50
    ```

>     Note that this command creates a VM with `50 GiB` disk size, 

Reviewer: Let's bump it up to 100GB

Author: Done!

    sudo curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc && \
    sudo chmod a+r /etc/apt/keyrings/docker.asc && \
    echo \
      "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \
      $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
      sudo tee /etc/apt/sources.list.d/docker.list > /dev/null && \
    sudo apt-get update
    ```

2. Install Docker

    ```bash
    sudo apt-get -y install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin && \
    sudo usermod -aG docker ${USER} && \
    newgrp docker
    ```

    You may follow [Docker documentation](https://docs.docker.com/engine/install/ubuntu/) 
    on details on installed Docker on Ubuntu.

3. Login to Docker

    <Tabs
     groupId="cloud"
     defaultValue="gcp"
     values={[
      { label: 'GCP', value: 'gcp', }
     ]
    }>
     <TabItem value="gcp">
        - Run the following command on the VM.
          ```bash
          gcloud auth login
          ```

        - Follow the on-screen prompts, it will display a URL that you need to copy-paste it

>           on the browser of your computer, _not_ the VM.

Reviewer: ```suggestion
          on the browser of your computer (_not_ the VM).
```

        - Follow the prompts on your browser, and login with an account that will provide you

>           with ccess to the GCR repository. If you are planning on _publishing_ to images you 

Reviewer: ```suggestion
          with access to the GCR repository. If you are planning on _publishing_ images you 
```

          to GCR. 

        - Configure Docker with your credentials.

          ```bash
          gcloud auth configure-docker
          ```

        You may refer to [this page](https://cloud.google.com/artifact-registry/docs/docker/authentication)
        for more details on configure Docker to access GCR.
     </TabItem>
    </Tabs>

## Checkout codebase {#checkout}

1. Clone the repository or its fork that contains the branch with the changes 
that you want to build the Docker images based-off. 

    ```shell
    git clone https://github.com/broadinstitute/gatk-sv && cd gatk-sv
    ```

2. Checkout the branch containing your changes.

    ```shell
    git checkout <BRANCH_NAME>
    ```

## Build and Publish Docker Images {#build}

In its minimal setup, you may use the following command to **build and publish** GATK-SV Docker images.

```shell
python3 scripts/docker/build_docker.py \
    --targets <IMAGES> \
    --image-tag <TAG> \
    --docker-repo <CONTAINER_REGISTRY>
```

The arguments are explained in the following.

> - [`--targets`](#targets);
> - [`--image-tag`](#tag);
> - [`--docker-repo`](#registry).

Reviewer: ```suggestion
- [`--targets`](#targets)
- [`--image-tag`](#tag)
- [`--docker-repo`](#registry)
```

  3. Run the script using `--base-git-commit` and `--current-git-commit` instead of `--targets`. 
  ```shell
  python scripts/docker/build_docker.py \
      --base-git-commit <BASE_SHA> \
      --current-git-commit <HEAD_SHA>
  ```
     
Please note that `--targets` and `--base-git-commit --current-git-commit` 
options are mutually exclusive. In other words, you can either manually specify 
images to rebuild, or let the script determine them automatically using commit SHAs; 
combining or avoiding both options is not currently supported.

:::info
Following the steps above, the script builds the specified Docker images 
_and all the images derived from them_. 
You may add the `--skip-dependent-images` flag to build only the explicitly specified images.
:::

### `--image-tag` {#tag}
 
You may use any naming convention for the Docker image 
[tags](https://docs.docker.com/engine/reference/commandline/tag/). 
GATK-SV Docker images are tagged using the following template 
(you may refer to [this section](./automated#args) for details).

```
[Date]-[Release Tag]-[Head SHA 8]
```

For example:

```

> 2023-07-28-v0.28.1-beta-e70dfbd7

Reviewer: ```suggestion
--image-tag 2023-07-28-v0.28.1-beta-e70dfbd7
```

```

### ` --docker-repo` {#registry}

If you are only testing GATK-SV Docker image build, 
you may skip this section and avoid providing `--docker-repo <registry>`.

> However, you need to push image to container registries,
> if you need images for WDL testing, or need to host the images on a container registry 

Reviewer: ```suggestion
However, if you need to push image to container registries, 
need images for WDL testing, or need to host the images on a container registry 
```
